libs/unity/library/Runtime/Scripts/Media/MicrophoneSource.cs

```diff
@@ -36,77 +32,55 @@ public class MicrophoneSource : AudioTrackSource
-                    // and check for permission again each time (see below why).
-                    _androidRecordAudioRequestPending = true;
-                    _androidRecordAudioRequestRetryUntilTime = Time.time + 300;
-
-                    // Display dialog requesting user permission. This will return immediately,
-                    // and unfortunately there's no good way to tell when this completes. As a rule
-                    // of thumb, application should lose focus, so check when focus resumes should
-                    // be sufficient without having to poll every frame.
-                    Permission.RequestUserPermission(Permission.Microphone);
-                }
-                return;
-            }
-#endif
 
-#if UNITY_WSA && !UNITY_EDITOR
-            // Request access to audio capture. The OS may show some popup dialog to the
-            // user to request permission. This will succeed only if the user approves it.
-            try
+            // Continue the task outside the Unity app context, in order to avoid deadlock
+            // if OnDisable waits on this task.
+            bool accessGranted = await RequestAccessAsync().ConfigureAwait(false);
+            if (!accessGranted)
             {
-                if (UnityEngine.WSA.Application.RunningOnUIThread())
-                {
-                    await RequestAccessAsync();
-                }
-                else
-                {
-                    UnityEngine.WSA.Application.InvokeOnUIThread(() => RequestAccessAsync(), waitUntilDone: true);
-                }
+                return null;
             }
-            catch (Exception ex)
-            {
-                // Log an error and prevent activation
-                Debug.LogError($"Audio access failure: {ex.Message}.");
-                this.enabled = false;
-                return;
-            }
-#endif
 
             var initConfig = new LocalAudioDeviceInitConfig
             {
                 AutoGainControl = _autoGainControl,
             };
             try
             {
-                AttachSource(await DeviceAudioTrackSource.CreateAsync(initConfig));
+                var createTask = DeviceAudioTrackSource.CreateAsync(initConfig);
+                // Continue the task outside the Unity app context, in order to avoid deadlock
+                // if OnDisable waits on this task.
+                return await createTask.ConfigureAwait(continueOnCapturedContext: false);
             }
             catch (Exception ex)
             {
                 Debug.LogError($"Failed to create device track source for {nameof(MicrophoneSource)} component '{name}'.");
                 Debug.LogException(ex, this);
-                return;
+                throw ex;
+            }
+        }
+
+        protected void Update()
+        {
+            var result = _initHelper.Result;
+            if (result != null)
+            {
+                AttachSource(result);
             }
```

Comment:
Polling every frame for a one-off thing :cry:

Reply:
So, we want to run `AttachSource` on the Unity app thread. The simplest thing would be to do it at the end of the initialization task, but that's setup so that it doesn't complete on the app thread (to avoid deadlocking with OnDisable). 

Another thing we could do is capturing the Unity `SynchronizationContext` and posting the call to `AttachSource` on it (which btw is probably a better way to run tasks on the app thread than our `WorkQueue`) at the end of the initialization task but _only_ if the component hasn't been disable yet, but this requires non-trivial synchronization. Also using the Unity `SynchronizationContext` directly is not well documented and there might be pitfalls.

There shouldn't be too many device sources in a single app, so I'd say let's leave it like this

Reply:
Good catch about the `SynchronizationContext`, but no idea if Unity supports that well nor at all. Let's leave as is for now if that works.